rfcs/0126-generic-worker-v2.md

```diff
@@ -0,0 +1,140 @@
+# RFC 126 - Megalodon worker (a.k.a. generic-worker v2)
+* Comments: [#126](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/126)
+* Proposed by: @walac
+
+# Summary
+
+The worker is the component of Taskcluster responsible to execute the
+job described in the task payload. It runs on a cloud instance or
+data center server and picks its jobs from the
+[queue](https://github.com/taskcluster/taskcluster-queue).
+
+Taskcluster projects currently provides two workers:
+
+* [docker-worker](https://github.com/taskcluster/docker-worker): executes
+Taskcluster jobs inside a docker container.
+
+* [generic-worker](https://github.com/taskcluster/generic-worker): executes
+Taskcluster jobs natively in the running machine.
+
+Although workers may run tasks in rather different environments, they share
+a fair amount of features, and it is reasonable to think they could share
+a common code base. Indeed this was the finality of
+[taskcluster-worker](https://github.com/taskcluster/taskcluster-worker),
+a worker aimed to support multiple execution environments. For reasons
+beyond the scope of this document, it failed to achieve this goal.
+
+# Motivation
+
+Both docker-worker and generic-worker were designed to accomplish a very
+specific task and do it well, but their architecture opens little room
+to support different execution environments, hence the proposal for a
+ground up worker.
+
+# Implementation
+
+Note: we assume we will implement the worker in  [Go](https://golang.org)
+programming language.
```

Comment:
I have zero knowledge of Rust, so I don't have an opinion here. `Go` is a simple language, it is built to manage concurrent server side, its tooling is pretty good (including [dependency managent](https://github.com/golang/dep) and has loads of stable library out there. Would be able to provide the technical advantages of `Rust` compared to `Go`, from a Taskcluster needs POV?

Reply:
I'm no fan of Go, and I would *love* to write Rust for work, but the idea of starting a new language along with a new implementation kind of scares me.  My experience with Rust has been that it lacks big swathes of features and is still very much oriented toward quick demos, rather than development of large, mature applications.  For example, last time I tried Async rust, it was a thing of beauty as long as you wrote everything as a nested function inside `main`, but once you left `main` the type inference broke down and you had to write out huge Future types or, in many cases, it was impossible to write the type.  Impl trait fixes this, but it'd be really frustrating to run into such issues in a high-priority project like this.